source/extensions/filters/listener/http_inspector/http_inspector.cc

```diff
@@ -0,0 +1,174 @@
+  ENVOY_LOG(trace, "http inspector: recv: {}", result.rc_);
+
+  if (result.rc_ == -1 && result.errno_ == EAGAIN) {
+    return;
+  } else if (result.rc_ < 0) {
+    config_->stats().read_error_.inc();
+    done(false);
+    return;
+  }
+
+  if (static_cast<uint64_t>(result.rc_) > read_) {
+    const uint8_t* data = buf_ + read_;
+    const size_t len = result.rc_ - read_;
+    read_ = result.rc_;
+    parseHttpHeader(absl::string_view(reinterpret_cast<const char*>(data), len));
+  } else {
+    done(false);
+  }
+}
+
+void Filter::parseHttpHeader(absl::string_view data) {
+  if (absl::StartsWith(data, Filter::HTTP2_CONNECTION_PREFACE)) {
+    ENVOY_LOG(trace, "http inspector: http2 connection preface found");
+    protocol_ = "HTTP/2";
+    done(true);
+  } else {
+    unsigned int id, sid = 0, spaces[2];
+    for (id = 0; id + 1 < data.length(); ++id) {
+      // The end of the request line.
+      if (data[id] == '\r' && data[id + 1] == '\n') {
+        break;
+      }
+
+      if (data[id] == ' ') {
+        if (sid < 2) {
+          spaces[sid++] = id;
+        } else {
+          sid = 0;
+          break;
+        }
+      }
+    }
+
+    if (sid != 2 || id + 1 == data.length()) {
+      ENVOY_LOG(trace, "http inspector: invalid http1x request line");
+      done(false);
+      return;
+    }
+
+    absl::string_view method(data.data(), spaces[0]);
+    // TODO(crazyxy): check request URI
+    absl::string_view request_uri(data.data() + spaces[0] + 1, spaces[1] - spaces[0] - 1);
+    absl::string_view http_version(data.data() + spaces[1] + 1, id - spaces[1] - 1);
+
+    if (httpMethods().count(method) == 0 || httpProtocols().count(http_version) == 0) {
+      ENVOY_LOG(trace, "http inspector: method: {} or protocol: {} not found", method,
+                http_version);
+      done(false);
+      return;
+    }
+
+    ENVOY_LOG(trace, "http inspector: method: {}, request uri: {}, protocol: {}", method,
+              request_uri, http_version);
+
+    protocol_ = http_version;
+    done(true);
+  }
+}
+
+void Filter::done(bool success) {
+  ENVOY_LOG(trace, "http inspector: done: {}", success);
+
+  if (success) {
+    absl::string_view protocol;
+    if (protocol_ == Http::Headers::get().ProtocolStrings.Http10String) {
+      config_->stats().http10_found_.inc();
+      protocol = "http/1.0";
+    } else if (protocol_ == Http::Headers::get().ProtocolStrings.Http11String) {
+      config_->stats().http11_found_.inc();
+      protocol = "http/1.1";
+    } else {
+      ASSERT(protocol_ == "HTTP/2");
+      config_->stats().http2_found_.inc();
+      protocol = "h2";
+    }
+
+    cb_->socket().setRequestedApplicationProtocols({protocol});
+  } else {
+    config_->stats().http_not_found_.inc();
+  }
+
+  file_event_.reset();
+  // Do not skip other listener filters.
+  cb_->continueFilterChain(true);
+}
+
+const absl::flat_hash_set<std::string>& Filter::httpMethods() const {
```

Comment:
IIRC HTTP spec doesn't limit methods to those? @PiotrSikora 

At least we have to include PATCH method here (RFC5789)

Reply:
Yes, there is a lot more (see official [HTTP Method Registry](https://www.iana.org/assignments/http-methods/http-methods.xhtml)), though even that registry is missing some popular but non-standardized methods (e.g. `PURGE`), so I'm not sure if restricting this to only well-known methods is justified.

Reply:
Since building a full set which contains both standard and non-standard methods is very difficult and we also check the http protocol later, maybe we remove the check of methods? Meanwhile, we can output a log if the method is not in the set. WDYT? @lizan @PiotrSikora

Reply:
Yeah, just make sure there is something with non-zero length.